File: apps/retro_hex_chat_web/lib/retro_hex_chat_web/components/ui/shell/menu_bar.ex
defmodule RetroHexChatWeb.Components.UI.MenuBar do
  @moduledoc """
  Screen-agnostic building blocks for a macOS-style textual menu bar.

  These primitives carry the DOM contract expected by `MenuBarHook`
  (`assets/js/hooks/ui/menu_bar_hook.js`) so it lives in exactly one place:

    * `menu_bar/1` — the `<nav role="menubar">` the hook mounts on.
    * `menu/1` — one top-level menu: a `[data-menubar-trigger]` and its
      `[data-menubar-dropdown]` panel as siblings under a single wrapper, the
      structure the hook relies on (`trigger.parentElement` → `querySelector`).
    * `submenu/1` — a nested panel inside a dropdown, on its own
      `[data-menubar-submenu]` / `[data-menubar-submenu-panel]` contract so the
      hook can open it without the parent dropdown's sweep closing it.

  A disabled menu grays its trigger and omits its dropdown entirely (the hook
  bails on `data-disabled="true"`). Dropdown items are supplied by the caller as
  `<li>` elements — compose `RetroHexChatWeb.Components.UI.ContextMenu`
  (`context_menu_item`/`context_menu_label`/`context_menu_separator`), which the
  hook closes on click.

  Content lives in the concrete bars that compose these primitives
  (`MenuBarApp` for chat); nothing here is screen-specific.
  """
  use RetroHexChatWeb.Component

  alias RetroHexChatWeb.Icons

  @doc """
  Renders the menu bar `<nav>` — the mount point for `MenuBarHook`.

  Pass `phx-hook="MenuBarHook"` (and any `id`) through the global attrs; the
  `menu/1` children go in the default slot.
  """
  attr :id, :string, default: "menubar"
  attr :testid, :string, default: "menu-bar"
  attr :class, :any, default: nil
  attr :rest, :global
  slot :inner_block, required: true

  @spec menu_bar(map()) :: Phoenix.LiveView.Rendered.t()
  def menu_bar(assigns) do
    ~H"""
    <nav
      id={@id}
      class={classes(["flex items-center shrink-0 select-none", @class])}
      role="menubar"
      data-menubar-root
      data-testid={@testid}
      {@rest}
    >
      {render_slot(@inner_block)}
    </nav>
    """
  end

  @doc """
  Renders one top-level menu: a trigger button plus its dropdown panel.

  Every menu carries a mandatory semantic icon (16×16, from
  `RetroHexChatWeb.Icons`) rendered before its label, so all menu bars read
  consistently across screens.

  When `disabled` is true the trigger is grayed and the dropdown is not
  rendered at all — matching the hook, which ignores disabled triggers.
  """
  attr :label, :string, required: true
  attr :disabled, :boolean, default: false
  attr :testid, :string, default: nil
  attr :class, :any, default: nil
  attr :trigger_class, :any, default: nil
  attr :label_class, :any, default: nil

  attr :offline_disabled, :boolean,
    default: false,
    doc: """
    Marks a menu the connection hook grays out while the socket is down. It is
    an attribute rather than a name the hook recognises because the label is
    translated: matching "File" only ever worked in English.
    """

  slot :inner_block, required: true, doc: "dropdown items (context_menu_* <li> elements)"
  slot :icon, required: true, doc: "semantic 16×16 icon rendered before the trigger label"

  @spec menu(map()) :: Phoenix.LiveView.Rendered.t()
  def menu(assigns) do
    ~H"""
    <div class={classes(["relative inline-flex", @class])}>
      <.menu_trigger
        label={@label}
        disabled={@disabled}
        offline_disabled={@offline_disabled}
        testid={@testid}
        class={@trigger_class}
        label_class={@label_class}
      >
        <:icon>{render_slot(@icon)}</:icon>
      </.menu_trigger>
      <.menu_dropdown :if={!@disabled}>
        {render_slot(@inner_block)}
      </.menu_dropdown>
    </div>
    """
  end

  @doc """
  Renders a nested submenu as one row inside a dropdown.

  The row looks and reads like a `context_menu_item`, plus a right-pointing
  chevron, and reveals its own panel of `<li>` items. Use it to group a family
  of related commands that would otherwise bloat the parent dropdown.

  The panel deliberately does NOT carry `data-menubar-dropdown`: the hook's
  `_closeAll`/`_openMenu` sweep every element with that attribute, so reusing it
  here would make a submenu open and close with its parent. It carries
  `data-menubar-submenu-panel` instead, and the hook drives it separately.

  On desktop the panel flies out to the right of the parent row. In the stacked
  (mobile) shell it expands inline instead — a flyout would be clipped by the
  mobile menu's `overflow: hidden` — which is CSS-only (`app-menu.css`); the
  open/close contract is identical on both.
  """
  attr :label, :string, required: true
  attr :testid, :string, default: nil
  attr :class, :any, default: nil
  slot :icon, required: true, doc: "14×14 icon SVG, matching context_menu_item"
  slot :inner_block, required: true, doc: "submenu items (context_menu_* <li> elements)"

  @spec submenu(map()) :: Phoenix.LiveView.Rendered.t()
  def submenu(assigns) do
    ~H"""
    <li
      class={classes(["menubar-submenu group/submenu relative list-none", @class])}
      data-menubar-submenu
      data-submenu-open="false"
      role="none"
    >
      <div
        class={
          classes([
            "flex items-center gap-retro-6 px-retro-16 py-2.5 md:py-[2px] min-h-[44px] md:min-h-0",
            "whitespace-nowrap text-sm md:text-xs cursor-pointer select-none",
            "hover:bg-selection-bg hover:text-selection-fg",
            "group-data-[submenu-open=true]/submenu:bg-selection-bg",
            "group-data-[submenu-open=true]/submenu:text-selection-fg"
          ])
        }
        data-menubar-submenu-trigger
        data-testid={@testid}
        aria-haspopup="true"
      >
        <span class="shrink-0 w-[14px] h-[14px] inline-flex items-center justify-center">
          {render_slot(@icon)}
        </span>
        <span class="flex-1">{@label}</span>
        <Icons.icon_chevron_right class="ml-retro-24 h-[14px] w-[14px] shrink-0" />
      </div>

      <div
        class="menubar-submenu__panel u-hidden absolute left-full top-0 min-w-[180px] p-[3px] bg-surface text-foreground shadow-retro-window z-dropdown"
        data-menubar-submenu-panel
      >
        <ul class="list-none m-0 p-retro-2">
          {render_slot(@inner_block)}
        </ul>
      </div>
    </li>
    """
  end

  @doc """
  Renders the stacked shell's icon rail — one button per menu, spanning the
  header.

  This is what a phone taps instead of the desktop's textual strip: every menu
  keeps a place of its own, and the space a lone hamburger left empty becomes
  the touch target. The engine reads `data-mobile-menu-open` to open the shared
  drawer already on that section, so the rail carries no dropdown of its own —
  pair it with `mobile_menu_drawer/1` over the same `sections`.
  """
  attr :sections, :list,
    required: true,
    doc: "%{id, label, icon_fn} maps, in the order the rail shows them"

  attr :class, :any, default: nil

  @spec mobile_menu_rail(map()) :: Phoenix.LiveView.Rendered.t()
  def mobile_menu_rail(assigns) do
    ~H"""
    <div
      class={classes(["app-menu-bar__mobile-rail", @class])}
      role="group"
      aria-label={dgettext("ui", "Menu")}
      data-testid="app-mobile-menu-rail"
    >
      <button
        :for={section <- @sections}
        type="button"
        class="app-menu-bar__mobile-rail-button"
        data-mobile-menu-open={section.id}
        data-active="false"
        data-testid={"app-mobile-menu-rail-#{section.id}"}
        aria-haspopup="true"
        aria-expanded="false"
        aria-label={section.label}
        title={section.label}
      >
        {apply(Icons, section.icon_fn, [%{class: "h-4 w-4"}])}
      </button>
    </div>
    """
  end

  @doc """
  Renders the one drawer every rail button opens, showing a section at a time.

  A drill-down rather than a stack of menus: a column of section tabs beside the
  items of whichever section is current. The rail, the tab column and this panel
  all name sections by the same id, which is how tapping any of the three lands
  on the same place.

  Pass one `:section` slot per entry in `sections`, matched by `id`.
  """
  attr :menu_id, :string, required: true
  attr :sections, :list, required: true, doc: "same list the paired rail renders"
  attr :active_section, :string, required: true
  attr :class, :any, default: nil

  slot :section, doc: "items for one section" do
    attr :id, :string, required: true
  end

  @spec mobile_menu_drawer(map()) :: Phoenix.LiveView.Rendered.t()
  def mobile_menu_drawer(assigns) do
    ~H"""
    <.menu
      class={classes(["app-menu-bar__mobile-menu", @class])}
      label={dgettext("ui", "Menu")}
      testid="app-mobile-menu-trigger"
      trigger_class="app-menu-bar__mobile-trigger shadow-retro-raised bg-surface h-8 w-10 justify-center border-0 px-0 py-0"
      label_class="sr-only"
    >
      <:icon><Icons.icon_btn_menu class="h-4 w-4" /></:icon>

      <li class="app-mobile-menu__shell" data-mobile-menu-root role="none">
        <div class="app-mobile-menu__categories" role="tablist" aria-label={dgettext("ui", "Menu")}>
          <button
            :for={section <- @sections}
            id={mobile_menu_category_id(@menu_id, section.id)}
            type="button"
            class="app-mobile-menu__category"
            data-mobile-menu-category={section.id}
            data-active={@active_section == section.id && "true"}
            data-testid={"app-mobile-menu-category-#{section.id}"}
            aria-selected={to_string(@active_section == section.id)}
            aria-controls={mobile_menu_panel_id(@menu_id, section.id)}
            role="tab"
          >
            <span class="app-mobile-menu__category-icon">
              {apply(Icons, section.icon_fn, [%{class: "h-[14px] w-[14px]"}])}
            </span>
            <span class="app-mobile-menu__category-label">{section.label}</span>
          </button>
        </div>

        <div class="app-mobile-menu__content">
          <section
            :for={panel <- @section}
            id={mobile_menu_panel_id(@menu_id, panel.id)}
            class={classes(["app-mobile-menu__panel", @active_section != panel.id && "u-hidden"])}
            data-mobile-menu-section={panel.id}
            data-testid={"app-mobile-menu-section-#{panel.id}"}
            aria-labelledby={mobile_menu_category_id(@menu_id, panel.id)}
            role="tabpanel"
          >
            <ul class="list-none m-0 p-0">{render_slot(panel)}</ul>
          </section>
        </div>
      </li>
    </.menu>
    """
  end

  # ── Private helpers ─────────────────────────────────

  defp mobile_menu_category_id(menu_id, section), do: "#{menu_id}-mobile-menu-category-#{section}"
  defp mobile_menu_panel_id(menu_id, section), do: "#{menu_id}-mobile-menu-section-#{section}"

  attr :label, :string, required: true
  attr :disabled, :boolean, default: false
  attr :offline_disabled, :boolean, default: false
  attr :testid, :string, default: nil
  attr :class, :any, default: nil
  attr :label_class, :any, default: nil
  slot :icon, required: true

  defp menu_trigger(assigns) do
    ~H"""
    <button
      type="button"
      class={
        classes([
          "inline-flex items-center gap-1 px-2 py-px text-sm border border-transparent whitespace-nowrap",
          if(@disabled,
            do: "text-muted-foreground cursor-default",
            else: "bg-transparent cursor-pointer hover:bg-accent"
          ),
          @class
        ])
      }
      data-menubar-trigger
      data-offline-disabled={if(@offline_disabled, do: "true", else: "false")}
      data-disabled={if(@disabled, do: "true", else: "false")}
      data-testid={@testid}
      aria-haspopup="true"
      aria-label={@label}
    >
      <span class={
        classes([
          "inline-flex h-4 w-4 shrink-0 items-center justify-center",
          @disabled && "opacity-50"
        ])
      }>
        {render_slot(@icon)}
      </span>
      <span class={@label_class} data-menubar-label>{@label}</span>
    </button>
    """
  end

  slot :inner_block, required: true

  defp menu_dropdown(assigns) do
    ~H"""
    <div
      class="u-hidden absolute top-full left-0 min-w-[180px] p-[3px] bg-surface shadow-retro-window z-dropdown"
      data-menubar-dropdown
      data-escape-guard
    >
      <ul class="list-none m-0 p-retro-2">
        {render_slot(@inner_block)}
      </ul>
    </div>
    """
  end
end

File: apps/retro_hex_chat_web/assets/js/lib/ui/menu_bar.js
/**
 * macOS-style menu bar — the engine, free of any framework.
 *
 * Owns every menu interaction: opening and hot-tracking the top-level
 * dropdowns, the flyout submenus, and the mobile rail that drives a single
 * shared drawer. It reads the DOM the server already rendered and decorates it,
 * which is what lets the same bar run on a public page with no LiveSocket at
 * all — the markup stands on its own and this is pure enhancement.
 *
 * Create one with `createMenuBar(el)` and drive it through `mount()` and
 * `destroy()`. The DOM contract (see `Components.UI.MenuBar`):
 *
 *   - `[data-menubar-trigger]`         opens the sibling dropdown under one wrapper
 *   - `[data-menubar-dropdown]`        the panel a trigger opens
 *   - `[data-menubar-submenu]`         a nested row, with its own panel
 *   - `[data-menubar-submenu-panel]`   that row's flyout — deliberately NOT a
 *                                      dropdown, so the parent sweep leaves it be
 *   - `[data-mobile-menu-open=<id>]`   rail button selecting a drawer section
 *   - `[data-mobile-menu-root]`        the drawer shell
 *   - `[data-mobile-menu-category=<id>]` section tab inside the drawer
 *   - `[data-mobile-menu-section=<id>]`  section panel inside the drawer
 *
 * Copy-selection items (`[data-menubar-copy-selection]`) are driven by
 * `copy_selection.js`, shared with the Start menu, which offers the same entry.
 *
 * Menus open on mousedown with preventDefault() so the composer — and the phone
 * keyboard with it — never loses focus to a menu.
 */
import { handleCopySelectionClick, refreshCopySelectionItems } from "./copy_selection";

const MenuBarCore = {
  mount() {
    this._activeMenu = null;
    this._mobileSection = null;

    this._onMouseDown = (e) => this._handleMouseDown(e);
    this._onMouseEnter = (e) => this._handleMouseEnter(e);
    this._onClick = (e) => this._handleClick(e);
    this._onOutside = (e) => {
      if (!this.el.contains(e.target)) this._closeAll();
    };
    this._onEscape = (e) => {
      if (e.key === "Escape") this._closeAll();
    };
    this._onForceClose = () => this._closeAll();

    this.el.addEventListener("mousedown", this._onMouseDown);
    this.el.addEventListener("mouseenter", this._onMouseEnter, true);
    this.el.addEventListener("click", this._onClick);
    this.el.addEventListener("menubar:close-all", this._onForceClose);
    document.addEventListener("mousedown", this._onOutside);
    document.addEventListener("keydown", this._onEscape);
  },

  destroy() {
    this.el.removeEventListener("mousedown", this._onMouseDown);
    this.el.removeEventListener("mouseenter", this._onMouseEnter, true);
    this.el.removeEventListener("click", this._onClick);
    this.el.removeEventListener("menubar:close-all", this._onForceClose);
    document.removeEventListener("mousedown", this._onOutside);
    document.removeEventListener("keydown", this._onEscape);
  },

  _handleMouseDown(e) {
    // The rail opens on mousedown like every other trigger, so the same
    // preventDefault keeps the focus (and the phone keyboard) on the composer.
    const railButton = e.target.closest("[data-mobile-menu-open]");
    if (railButton) {
      e.preventDefault();
      this._openMobileSection(railButton.dataset.mobileMenuOpen);
      return;
    }

    const trigger = e.target.closest("[data-menubar-trigger]");
    if (!trigger) return;
    e.preventDefault();

    if (trigger.dataset.disabled === "true") return;

    const menu = trigger.parentElement;
    if (this._activeMenu === menu) {
      this._closeAll();
    } else {
      this._openMenu(menu);
    }
  },

  // Hot tracking: hover switches menus when one is already open.
  _handleMouseEnter(e) {
    if (!this._activeMenu) return;

    // Inside an open dropdown, hovering a row drives the submenus rather
    // than the top-level menus.
    const row = e.target.closest?.("[data-menubar-dropdown] li");
    if (row) {
      this._trackSubmenuHover(row);
      return;
    }

    const trigger = e.target.closest("[data-menubar-trigger]");
    if (!trigger || trigger.dataset.disabled === "true") return;
    const menu = trigger.parentElement;
    if (menu !== this._activeMenu) this._openMenu(menu);
  },

  _handleClick(e) {
    const mobileCategory = e.target.closest("[data-mobile-menu-category]");

    if (mobileCategory) {
      e.preventDefault();
      e.stopPropagation();
      this._activateMobileMenuCategory(mobileCategory);
      return;
    }

    if (handleCopySelectionClick(e)) {
      this._closeAll();
      return;
    }

    // A submenu row is an <li> inside a dropdown, so it would otherwise hit
    // the close-everything branch below. Toggle it instead.
    const submenuTrigger = e.target.closest("[data-menubar-submenu-trigger]");

    if (submenuTrigger) {
      e.preventDefault();
      e.stopPropagation();

      const submenu = submenuTrigger.closest("[data-menubar-submenu]");

      if (submenu) {
        // A pointer hovers a row on its way to clicking it, so by the time
        // the click lands the hover has already opened the flyout. Toggling
        // on every click therefore closed it again, which put the submenu out
        // of reach of a mouse entirely. Only a click on a submenu that a
        // previous *click* opened counts as "put it away".
        const openedByClick =
          submenu.dataset.submenuOpen === "true" && submenu.dataset.submenuVia === "click";

        if (openedByClick) {
          this._setSubmenu(submenu, false);
        } else {
          this._openSubmenu(submenu, "click");
        }
      }

      return;
    }

    if (e.target.closest("[data-menubar-dropdown] li")) this._closeAll();
  },

  _openMenu(menu) {
    this._closeAll();
    const dropdown = menu.querySelector("[data-menubar-dropdown]");
    if (!dropdown) return;
    refreshCopySelectionItems(menu);
    dropdown.classList.remove("u-hidden");
    const trigger = menu.querySelector("[data-menubar-trigger]");
    if (trigger) {
      trigger.classList.add("bg-primary", "text-primary-foreground");
      trigger.classList.remove("hover:bg-accent");
    }
    this._activeMenu = menu;
  },

  // Opens the shared mobile dropdown already showing `section`. The rail has no
  // dropdown of its own: every button drives the one panel the mobile menu
  // renders, which is why the same button closes it and a different one only
  // swaps the section under an already-open drawer.
  _openMobileSection(section) {
    const root = this.el.querySelector("[data-mobile-menu-root]");
    if (!root) return;

    const menu = root.closest("[data-menubar-dropdown]")?.parentElement;
    if (!menu) return;

    if (this._activeMenu === menu && this._mobileSection === section) {
      this._closeAll();
      return;
    }

    if (this._activeMenu !== menu) this._openMenu(menu);

    const category = root.querySelector(`[data-mobile-menu-category="${section}"]`);
    if (category) this._activateMobileMenuCategory(category);
  },

  // Mirrors the open section onto the rail, so the button reads as pressed the
  // way an open menu's trigger does on the desktop. `null` clears every button.
  _syncMobileRail(section) {
    this.el.querySelectorAll("[data-mobile-menu-open]").forEach((button) => {
      const active = button.dataset.mobileMenuOpen === section;
      button.dataset.active = active ? "true" : "false";
      button.setAttribute("aria-expanded", active ? "true" : "false");
    });
  },

  _closeAll() {
    this._mobileSection = null;
    this._syncMobileRail(null);
    this.el.querySelectorAll("[data-menubar-dropdown]").forEach((d) => d.classList.add("u-hidden"));
    this.el
      .querySelectorAll("[data-menubar-submenu]")
      .forEach((submenu) => this._setSubmenu(submenu, false));
    this.el.querySelectorAll("[data-menubar-trigger]").forEach((t) => {
      t.classList.remove("bg-primary", "text-primary-foreground");
      if (t.dataset.disabled !== "true") {
        t.classList.add("hover:bg-accent");
      }
    });
    this._activeMenu = null;
  },

  // Hovering a submenu row opens it; hovering any sibling row closes the
  // submenus of that dropdown, so at most one flyout is ever showing.
  _trackSubmenuHover(row) {
    const submenu = row.closest("[data-menubar-submenu]");

    if (submenu) {
      this._openSubmenu(submenu);
      return;
    }

    const dropdown = row.closest("[data-menubar-dropdown]");
    if (dropdown) this._closeSiblingSubmenus(dropdown, null);
  },

  _openSubmenu(submenu, via = "hover") {
    const dropdown = submenu.closest("[data-menubar-dropdown]");
    if (dropdown) this._closeSiblingSubmenus(dropdown, submenu);

    // Hovering inside an already-clicked submenu must not downgrade it back to
    // "hover", or the click that put it away would open it instead.
    const alreadyClicked =
      submenu.dataset.submenuOpen === "true" && submenu.dataset.submenuVia === "click";

    this._setSubmenu(submenu, true, alreadyClicked ? "click" : via);
  },

  _closeSiblingSubmenus(dropdown, except) {
    dropdown.querySelectorAll("[data-menubar-submenu]").forEach((submenu) => {
      if (submenu !== except) this._setSubmenu(submenu, false);
    });
  },

  // `via` records what opened the flyout — a hover or a deliberate click — so
  // the click handler can tell "the pointer just arrived here" from "the reader
  // is clicking this a second time to put it away".
  _setSubmenu(submenu, open, via = "hover") {
    submenu.dataset.submenuOpen = open ? "true" : "false";

    if (open) {
      submenu.dataset.submenuVia = via;
    } else {
      delete submenu.dataset.submenuVia;
    }

    const panel = submenu.querySelector("[data-menubar-submenu-panel]");
    if (panel) panel.classList.toggle("u-hidden", !open);
  },

  _activateMobileMenuCategory(category) {
    const root = category.closest("[data-mobile-menu-root]");
    if (!root) return;

    const section = category.dataset.mobileMenuCategory;
    root.querySelectorAll("[data-mobile-menu-category]").forEach((button) => {
      const active = button.dataset.mobileMenuCategory === section;
      button.dataset.active = active ? "true" : "false";
      button.setAttribute("aria-selected", active ? "true" : "false");
    });
    root.querySelectorAll("[data-mobile-menu-section]").forEach((panel) => {
      panel.classList.toggle("u-hidden", panel.dataset.mobileMenuSection !== section);
    });

    // The category column inside the drawer switches sections too, so the rail
    // learns the section from here rather than from the tap that opened it.
    this._mobileSection = section;
    this._syncMobileRail(section);
  },
};

/** Builds a menu bar over `el` (a `<nav role="menubar">`). */
export function createMenuBar(el) {
  const bar = Object.create(MenuBarCore);
  bar.el = el;
  return bar;
}

export default MenuBarCore;
